qiskit_ode/signals/signals.py

```diff
@@ -11,216 +11,262 @@
 
 
-class BaseSignal(ABC):
-    """Base class for a time-dependent mixed signal."""
+class Signal:
+    r"""General signal class representing a function of the form:
 
-    def __init__(self, name: str = None):
-        """Init function."""
-        self._name = name
-        self._carrier_freq = Array(0.0)
-        self._phase = Array(0.0)
+    .. math::
+        Re[f(t)e^{i (2 \pi \nu t + \phi)}]
+                = Re[f(t)]\cos(2 \pi \nu t + \phi) - Im[f(t)]\sin(2 \pi \nu t + \phi),
 
-    @property
-    def carrier_freq(self):
-        """Return the carrier frequency of the signal."""
-        return self._carrier_freq
+    where
 
-    @property
-    def phase(self):
-        """Return the phase of the Signal."""
-        return self._phase
+    - :math:`f(t)` is the envelope function.
+    - :math:`\nu` is the carrier frequency.
+    - :math:`\phi` is the phase.
 
-    @abstractmethod
-    def conjugate(self):
-        """Return a new signal that is the complex conjugate of self."""
+    The envelope function can be specified either as a constant numeric value
+    (indicating a constant function), or as a complex-valued callable,
+    and the frequency and phase must be real.
 
-    @abstractmethod
-    def envelope_value(self, t: float = 0.0) -> complex:
-        """Evaluates the envelope at time t."""
+    Notes:
 
-    @abstractmethod
-    def value(self, t: float = 0.0) -> float:
-        """Return the value of the signal at time t."""
+    - If the envelope is specified as a constant, the real part is automatically taken.
+    - If the envelope is passed as a callable, this class assumes that the function is
+      vectorized: i.e. if called with an array of time values, it will return an array of the
+      same shape via entry-wise application of the function. If it isn't, it can be vectorized
+      automatically by calling ``numpy.vectorize``\, or, if using JAX, by
+      calling ``jax.numpy.vectorize``\. E.g.:
 
-    @property
-    def name(self) -> str:
-        """Return the name of the signal."""
-        return self._name
+    .. code-block:: python
 
-    def to_pwc(self, dt: float, n_samples: int, start_time: float = 0.0) -> "PiecewiseConstant":
+        vectorized_func = np.vectorize(non_vectorized_func)
+        signal = Signal(envelope=vectorized_func, carrier_freq=5.)
+    """
+
+    def __init__(
+        self,
+        envelope: Union[Callable, complex, float, int],
+        carrier_freq: float = 0.0,
+        phase: float = 0.0,
+        name: Optional[str] = None,
+    ):
         """
-        Converts a signal to a piecewise constant signal.
+        Initializes a signal given by an envelope and a carrier.
 
         Args:
-            dt: Time increment to use.
-            n_samples: number of steps to resample with.
-            start_time: start time from which to resample.
-
-        Returns:
-            A piecewiseConstant signal.
+            envelope: Envelope function of the signal, must be vectorized.
```

Comment:
From the docstring I would expect init to raise an error if the envelop is not vectorized but I cannot find one. Why?

Reply:
I don't believe it's possible to check for this at instantiation. I think the only way to really find out that something _isn't_ vectorized is if it fails to produce arrays of the correct shape on some input. I'm not sure what we could do here aside from adding checks in every signal evaluation function that the input shape = output shape, but this seems like unnecessary overhead. I'm tempted to not worry about this for now: if something isn't vectorized I believe it should result in pretty clear errors (of course, I could be wrong).